Eio.Executor_pool

SGrondin opened the pull request:
This is a continuation of #584. After discussing the design, we've agreed to first merge a simple and minimalistic version. I will be adding barriers, job weights, etc., in an upcoming version 2.

SGrondin commented:
The only open question remaining is whether is check the Switch's status when submitting jobs. I'm thinking it could be worth it to avoid silent deadlocks when the user does something funky with their lifetimes.

SGrondin commented on lib_eio/executor_pool.ml:
```diff
@@ -0,0 +1,75 @@
+  instance
+
+let enqueue { sw; stream } f =
+  Switch.check sw;
```
@talex5 I believe `Switch.check` is not thread safe. Is there a safe way to accomplish this or should I just let the users who do strange things with their lifetimes risk deadlock?

SGrondin commented on lib_eio/executor_pool.ml:
```diff
@@ -0,0 +1,76 @@
+  instance
+
+let enqueue { stream; cancelled } f =
+  if Atomic.get cancelled then raise (Failure "Executor_pool: Switch cancelled");
```
I know this is somewhat overkill, but without it users doing truly horrifying things with their timelines can end up in a (well deserved) deadlocked. 😆

talex5 commented:
I experimented with some changes here: https://github.com/SGrondin/eio/compare/executorpool...talex5:eio:executorpool?expand=1

Needs a bit more thinking about, but this should ensure that jobs are always rejected if the pool shuts down, though it does add some cost to the fast-path.

SGrondin commented:
I like this a lot. Compared to the previous version it guarantees that even egregious misuse never leads to a deadlock at the cost of complexity and a small overhead. The previous version did a "best effort" to avoid deadlocks in such cases. I'll merge it into my branch and play with it a bit.

SGrondin commented:
I added more information to the mli, fixed a typo, and removed a few `Fiber.yield ()` from the tests (fewer are needed now that we don't use `Fiber.n_any`). The commit is [here](https://github.com/SGrondin/eio/commit/03bfde4fb5a589f3b9fb61c762b55d1daba0779f). I then recombined all those commits into one. I'm looking at `Stream.close` now.

SGrondin commented:
@talex5 I've made a small change in the API that I believe makes the Executor_pool more powerful without significantly increasing the complexity (both of the code and to the user). It's getting late in the design cycle to make this change, but I think you'll agree with it.

talex5 commented:
What about fixing the domain capacity at 1 and having users say what fraction of the CPU they expect to use? The capacity seems arbitrary and might be hard to pick if you're sharing one pool with several parts of your program.

SGrondin commented:
Yes, I went back and forth on that one. I'll switch it over 👍

talex5 reviewed (approved):
I pushed a couple of minor changes (cope with `weight:nan` and use `invalid_arg`). Looks good to merge now!

SGrondin commented on lib_eio/executor_pool.mli:
```diff
@@ -0,0 +1,58 @@
+    when the switch finishes, all domain workers and running jobs are cancelled.
+
+    @param domain_count The number of domain workers to create.
+                        This number should not exceed the number of CPU cores on your system.
```
Done, I added a note.

talex5 commented:
Thanks!

(I added a final minor fix to ensure the client's promise is resolved if the worker is cancelled just as it accepts a job.)

yawaramin commented on lib_eio/executor_pool.ml:
```diff
@@ -0,0 +1,74 @@
+
+let enqueue { queue } ~weight fn =
+  if not (weight >= 0. && weight <= 1.) (* Handles NaN *)
+  then Fmt.invalid_arg "Executor_pool: weight not >= 0.0 && <= 1.0" weight
```
Should this be:

```ocaml
Fmt.invalid_arg "Executor_pool: weight not >= 0.0 && <= 1.0: %f" weight
```

?
